Implement Mustache templates for messages

marcins opened the pull request:
A number of people in issue #12  have asked for the ability to customise the messages, such as adding build parameters etc.  By using Mustache templates for build messages this can now be done.

This is quite a large change, and probably not as heavily tested as it should be. It probably needs some unit tests too.  I am also not a Java developer, but "know" the language, so I might not have done everything optimally.

The README has been updates with information as to how the templates work, and the defaults should keep the messages the same as they were pre-Mustache.

This doesn't actually solve the original issue as it doesn't add the git branch name, and if the branch name isn't available through the accessible instance variables then it won't, I haven't checked.

It's probably a good start though!

Commit:
Add IDEA files to .gitignore

Commit:
Update Jenkins version and add Mustache in POM

Updated to use Jenkins 1.500 and added the Java implementation
of Mustache from https://github.com/spullara/mustache.java

Commit:
Use Mustache templates to generate notifications

The plugin will now use Mustache template from configuration when
creating HipChat notifications. This allows you to use any values
available in the build / project configuration, including your own
custom parameters.

Commit:
Update version to 0.1.3

Commit:
Add link to completed message

Commit:
Update plugin version past what's on the plugin repo

Commit:
Update README with info about using Mustache

Commit:
Fix empty message issues

When there is no config set there was an empty message being sent,
instead make sure the defaults are used.

Commit:
Update README with custom var info

Commit:
Remove MessageBuilder

MessageBuilder is no longer required since we now generate messages
using Mustache.

Commit:
Make sure "Cause" is available to Mustache

And some minor fixes to ensure there's no doubling up of "Started"
text.

Commit:
Fix notifier so that is uses job properties

Previous code was incorrectly getting message templates from
the global notifier instead of using the JobProperties object

Commit:
Simplify the "started" template

Use a single "trigger" string that's the same as the non-Mustache
version's string instead of separate cause & changes strings which
could cause duplications.

Commit:
Clean up config panels and add help files

Commit:
Readme update

marcins committed:
Fix README indentation

marcins commented:
If anyone wants to try the plugin with Mustache notification templates I have a snapshot HPI up in Dropbox:

https://dl.dropbox.com/u/5336861/hipchat.hpi

YMMV, disclaimer, etc, etc :-)  

Also check the [updated README](https://github.com/marcins/jenkins-hipchat-plugin/tree/feature-mustache-messages)

Commit:
Merge branch 'feature-mustache-messages' of https://github.com/marcins/jenkins-hipchat-plugin into feature-mustache-messages

Commit:
Update default completion msg to allow html

If the build fails then the completion message may
contain HTML. This means the default message needs to
triple 'stache "status" so that the correct output is generated.

rdohms commented:
Any progress on accepting this? I would really like to add the description to my messages since i have "per PR" builds the number just becomes obsolete.

marcins commented:
@rdohms you can always run my snapshot HPI instead of the "official" one if you want the functionality now?  Actually I should build a new one with the last 2 fixes in there, don't have time right now though.

Mange commented:
I am also interested in this.

phil-schneider commented:
Nice work.  Please add to master branch

Commit:
Use StringBuilder properly in getOpenLink

Inefficient use of concatenated strings with String builder when creating the link to the build job.

Commit:
Use StringUtils to check for blank strings

Style fix to check for blank strings based on feedback.

marcins commented:
For all those that are interested, I've posted the latest snapshot HPI with these changes (based on  at:

https://dl.dropbox.com/u/5336861/hipchat.hpi

I've been using this in production since I wrote it 2 months ago and haven't experienced any issues, although we do only have a small installation.

phil-schneider commented:
I have another questions. Does your template replaces any parameter that are set by the build. Default parameter like $BUILD_ID and also custom parameter from a parameteriezed build?

marcins commented:
@phil-schneider - have a look at the [README in my feature branch](https://github.com/marcins/jenkins-hipchat-plugin/blob/feature-mustache-messages/README.markdown), it gives examples of how to access stuff about the build from the mustache templates:

iamamused commented:
I installed this fork from the compiled https://dl.dropbox.com/u/5336861/hipchat.hpi link but for some reason I can't seem to customize the template. If I change either the global or project completion template I get an empty message (I just tried adding the word _test_ into the template).

```
Apr 8, 2013 12:59:44 PM jenkins.plugins.hipchat.HipChatNotifier$HipChatJobProperty prebuild
INFO: Invoking Started...
Apr 8, 2013 12:59:44 PM jenkins.plugins.hipchat.ActiveNotifier getChanges
INFO: Empty change...
Apr 8, 2013 12:59:44 PM jenkins.plugins.hipchat.StandardHipChatService publish
INFO: Posting: Jenkins to 172833: Feature Test - #21: Started by user anonymous (<a href='/job/Feature%20Test/21/'>Open</a>) green
Apr 8, 2013 12:59:45 PM hudson.model.Run execute
INFO: Feature Test #21 main build action completed: SUCCESS
Apr 8, 2013 1:00:00 PM jenkins.plugins.hipchat.StandardHipChatService publish
INFO: Posting: Jenkins to 172833:  green
```

Notice the last line is missing the message.

marcins commented:
Interesting.. I'll look into it.

On Tue, Apr 9, 2013 at 3:14 AM, Jeffrey Sambells
notifications@github.comwrote:

> I installed this fork from the compiled
> https://dl.dropbox.com/u/5336861/hipchat.hpi link but for some reason I
> can't seem to customize the template. If I change either the global or
> project completion template I get an empty message (I just tried adding the
> work test into the template.
> 
> Apr 8, 2013 12:59:44 PM jenkins.plugins.hipchat.HipChatNotifier$HipChatJobProperty prebuild
> INFO: Invoking Started...
> Apr 8, 2013 12:59:44 PM jenkins.plugins.hipchat.ActiveNotifier getChanges
> INFO: Empty change...
> Apr 8, 2013 12:59:44 PM jenkins.plugins.hipchat.StandardHipChatService publish
> INFO: Posting: Jenkins to 172833: Feature Test - #21: Started by user anonymous (<a href='/job/Feature%20Test/21/'>Open</a>) green
> Apr 8, 2013 12:59:45 PM hudson.model.Run execute
> INFO: Feature Test #21 main build action completed: SUCCESS
> Apr 8, 2013 1:00:00 PM jenkins.plugins.hipchat.StandardHipChatService publish
> INFO: Posting: Jenkins to 172833:  green
> 
> Notice the last line is messing the message.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jlewallen/jenkins-hipchat-plugin/pull/32#issuecomment-16064297
> .

## 

M.

iamamused commented:
Actually it only seem sto be blank if I modify the project specific template. If I modify the global one it seems to ignore my settings and uses the default one.

iamamused commented:
Also noticed that the global Build Server Url isn't applied in the Open link (everything lis linking as if it's a local file. It's like it's not reading my configuration. I'm running the latest version of Jenkins.

marcins commented:
From all those reports it looks like there's just an issue with applying
the global config then? Although I'm pretty sure my production install is
only using global config and it is working.

Will try and reproduce when I have some time (maybe add some unit tests!)

On Tue, Apr 9, 2013 at 11:13 AM, Jeffrey Sambells
notifications@github.comwrote:

> Also noticed that the global Build Server Url isn't applied in the Open
> link (everything lis linking as if it's a local file. It's like it's not
> reading my configuration. I'm running the latest version of Jenkins.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jlewallen/jenkins-hipchat-plugin/pull/32#issuecomment-16088805
> .

## 

M.

iamamused commented:
Ya, possibly. Playing with it some more it seems editing the global config doesn't actually update the configs in jobs/ until after I delete and then re-added the hipchat phase. That fixed the Build Server URL but the custom messages still didn't work.

marcins commented:
I think this was an issue I ran into with the plugin before I made my
changes - the global config is only loaded once when the plugin is
initialised (so either when you first add it, or when the server starts
up), it's not reloaded when it's changed. Job config is reloaded everytime.

On Tue, Apr 9, 2013 at 11:45 AM, Jeffrey Sambells
notifications@github.comwrote:

> Ya, possibly. Playing with it some more it seems editing the global config
> doesn't actually update the configs in jobs/ until after I delete and then
> re-added the hipchat phase. That fixed the Build Server URL but the custom
> messages still didn't work.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jlewallen/jenkins-hipchat-plugin/pull/32#issuecomment-16089697
> .

## 

M.

iamamused commented:
ok, still can't figure out why the templates don't work though. The global one is ignored and the local changes just cause it to be blank. Here's my test job config if you can shed any insight:

```
</project>SpeakFeels-Mac-mini:Feature Test jenkins$ cat config.xml 
<?xml version='1.0' encoding='UTF-8'?>
<project>
  <actions/>
  <description>This is just a test project to play with all the settings in Jenkins. Please ignore any output.</description>
  <keepDependencies>false</keepDependencies>
  <properties>
    <jenkins.plugins.hipchat.HipChatNotifier_-HipChatJobProperty plugin="hipchat@0.1.5-SNAPSHOT">
      <room></room>
      <startNotification>true</startNotification>
      <notifySuccess>true</notifySuccess>
      <notifyAborted>true</notifyAborted>
      <notifyNotBuilt>true</notifyNotBuilt>
      <notifyUnstable>true</notifyUnstable>
      <notifyFailure>true</notifyFailure>
      <messageTemplateStarted>This is a test template for start</messageTemplateStarted>
      <messageTemplateCompleted>This is a test template for complete</messageTemplateCompleted>
      <messageTemplateSuffix></messageTemplateSuffix>
    </jenkins.plugins.hipchat.HipChatNotifier_-HipChatJobProperty>
  </properties>
  <scm class="hudson.scm.NullSCM"/>
  <canRoam>true</canRoam>
  <disabled>false</disabled>
  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
  <triggers class="vector"/>
  <concurrentBuild>false</concurrentBuild>
  <builders>
    <hudson.tasks.Shell>
      <command>env;</command>
    </hudson.tasks.Shell>
  </builders>
  <publishers>
    <com.zanox.hudson.plugins.FTPPublisher plugin="ftppublisher@1.2">
      <siteName>example.com</siteName>
      <entries>
        <com.zanox.hudson.plugins.Entry>
          <filePath>$REMOTE_DIR/build</filePath>
          <sourceFile>upload/*/**</sourceFile>
        </com.zanox.hudson.plugins.Entry>
      </entries>
      <useTimestamps>false</useTimestamps>
      <flatten>true</flatten>
      <skip>false</skip>
    </com.zanox.hudson.plugins.FTPPublisher>
    <jenkins.plugins.hipchat.HipChatNotifier plugin="hipchat@0.1.5-SNAPSHOT">
      <authToken>...removed...</authToken>
      <buildServerUrl>http://example.com/</buildServerUrl>
      <room>172833</room>
      <sendAs>Jenkins</sendAs>
      <messageTemplateStarted>{{build.project.displayName}} - {{build.displayName}}: {{trigger}} {{{link}}} test</messageTemplateStarted>
      <messageTemplateCompleted>{{build.project.displayName}} - {{build.displayName}}: {{status}} after {{build.durationString}} test</messageTemplateCompleted>
      <messageTemplateSuffix></messageTemplateSuffix>
    </jenkins.plugins.hipchat.HipChatNotifier>
  </publishers>
  <buildWrappers>
    <EnvInjectBuildWrapper plugin="envinject@1.83">
      <info>
        <propertiesContent>REMOTE_DIR=feature-test</propertiesContent>
        <loadFilesFromMaster>false</loadFilesFromMaster>
      </info>
    </EnvInjectBuildWrapper>
  </buildWrappers>
</project>
```

drubin commented:
@marcins I just tested this. It seems that it works perfectly except for setting a custom template per project. When doing that (even the default one, or just plain text) it seems to not send the notification. 

Using the custom suffix seems to work. 

Also I would suggest adding some sort of logging so we can see the output in the build script (currently there is NO way to figure out why the messages aren't being sent if the template is wrong)

kellybyrd commented:
There seem to be three (this one, #38, #42) pull requests all about customizing the sent message. This one seems most flexible and could probably cover the use cases of the other two, but I worry about it being able to cover the dead simple: "I just want to change the text string a little bit" case without someone having to read the Mustache docs. How do the other maintainers feel?
